src/material-experimental/form-field/testing/form-field-harness.ts

```diff
@@ -0,0 +1,231 @@
+    // TODO(reviewer): not sure what we should do here. We need to construct a combined
+    // query for all possible controls. This is necessary because if a form-control contains
+    // multiple possible control candidates (i.e. a chip-list with matInput). We need to
+    // find the control that matches first (to simulate the behavior of "@ContentChild").
+    // The question is, how we can achieve this in a public manner. Should we expose these
+    // functions? seems reasonable to me.
+    const elementCandidates: TestElement[] =
+        (await (environment as any).getAllRawElements(this._controlHarnessSelector))
+            .map((rawElement: any) => (environment as any).createTestElement(rawElement));
+
+    if (!elementCandidates.length) {
+      return null;
+    }
+
+    const controlElement = elementCandidates[0];
+    const [isInput, isSelect] = await Promise.all([
+      controlElement.matchesSelector(MatInputHarness.hostSelector),
+      controlElement.matchesSelector(MatSelectHarness.hostSelector),
+    ]);
+
+    if (isInput) {
+      return (environment as any).createComponentHarness(MatInputHarness, controlElement);
+    } else if (isSelect) {
+      return (environment as any).createComponentHarness(MatSelectHarness, controlElement);
+    }
+    // This error will be reported if we add new harnesses to the form-field control
+    // selector but forget checking for it here.
+    throw Error('Found control of form-field but could not create harness.');
+  }
+
+  /** Whether the form-field has a label. */
+  async hasLabel(): Promise<boolean> {
+    return (await this.host()).hasClass('mat-form-field-has-label');
+  }
+
+  /** Gets the label of the form-field. */
+  async getLabel(): Promise<string|null> {
+    const labelEl = await this._label();
+    return labelEl ? labelEl.text() : null;
+  }
+
+  /** Whether the form-field has a floating label. */
+  async hasFloatingLabel(): Promise<boolean> {
+    return (await this.host()).hasClass('mat-form-field-can-float');
+  }
+
+  /** Whether the label is currently floating. */
+  async isLabelFloating(): Promise<boolean> {
+    return (await this.host()).hasClass('mat-form-field-should-float');
+  }
+
+  /** Whether the form-field is disabled. */
+  async isDisabled(): Promise<boolean> {
+    return (await this.host()).hasClass('mat-form-field-disabled');
+  }
+
+  /** Whether the form-field is currently autofilled. */
+  async isAutofilled(): Promise<boolean> {
+    return (await this.host()).hasClass('mat-form-field-autofilled');
+  }
+
+  /** Gets the theme color of the form-field. */
+  async getThemeColor(): Promise<'primary'|'accent'|'warn'> {
+    const hostEl = await this.host();
+    const [isAccent, isWarn] =
+        await Promise.all([hostEl.hasClass('mat-accent'), hostEl.hasClass('mat-warn')]);
+    if (isAccent) {
+      return 'accent';
+    } else if (isWarn) {
+      return 'warn';
+    }
+    return 'primary';
```

Comment:
A bit more concise?
```ts
const hostClass = await (await this.host()).getAttribute('class');
for (const palette of ['accent', 'warn']) {
  if (hostClass.includes(palette)) {
    return palette;
  }
}

return 'primary';
```

Reply:
Sounds like a good idea.

Reply:
I'll actually keep it like that as otherwise we'd need to handle more things (i.e. spaces to separate classes in the  `class` attribute). The current implementation is quite reasonable but just looks weird due to the Promise all for concurrency.

Reply:
You could maybe use `getProperty('classList')` for this, now that we have separate methods for attributes and properties

Reply:
I don't follow what you mean by handing more things. Checking for the absence/presence of "accent" or "warn" is enough, no?

Reply:
My thinking is that `class` can contain more classes and that just checking using `includes` is not working 100%. e.g. class attribute could be: `custom-mat-accent mat-primary-field`.

So we need to ensure that there is a space before/after the class we check for, or that it is at the start/end of the `class` attribute value. or just split the `class` attribute by space.. Surely these cases are extremely rare but I don't think there is a benefit in making these things less robust if we can just simply use the public robust `hasClass` methods.

The current implementation basically checks for existence of the theme classes. It just uses the intended APIs but unfortunately looks kind of dirty (to me at least)